error: package exists in another module: jdk.compiler

While running ASHE in dryrun mode, we are seeing this error message in **many** places, not just with this one specific case.

INFO  edu.njit.jerse.ashe.services.SpeciminTool - Compiling minimized files in directory: /my/temp/dir/with/minimized/files
INFO  edu.njit.jerse.ashe.services.SpeciminTool - Compiling Java files:
javac /my/temp/dir/with/minimized/files/org/plumelib/util/UniqueId.java /my/temp/dir/with/minimized/files/org/checkerframework/dataflow/analysis/AbstractValue.java /my/temp/dir/with/minimized/files/org/checkerframework/dataflow/analysis/AnalysisResult.java /my/temp/dir/with/minimized/files/org/checkerframework/dataflow/analysis/Store.java /my/temp/dir/with/minimized/files/org/checkerframework/dataflow/cfg/node/Node.java /my/temp/dir/with/minimized/files/com/sun/source/tree/Tree.java
00:27:29.288 [main] INFO  edu.njit.jerse.ashe.services.SpeciminTool - 
Specimin output:
```
/my/temp/dir/with/minimized/files/com/sun/source/tree/Tree.java:1: error: package exists in another module: jdk.compiler
package com.sun.source.tree;
```
^
1 error

--- a/src/main/java/org/checkerframework/specimin/InheritancePreserveVisitor.java
+++ b/src/main/java/org/checkerframework/specimin/InheritancePreserveVisitor.java
@@ -98,7 +98,7 @@ public class InheritancePreserveVisitor extends ModifierVisitor<Void> {
       for (ClassOrInterfaceType implementedType : decl.getImplementedTypes()) {
         try {
           String interfacename = implementedType.resolve().describe();
-          if (interfacename.startsWith("java.")) {
+          if (JavaLangUtils.inJdkPackage(interfacename)) {
             // Avoid keeping implementations of java.* classes, because those
             // would require us to actually implement them (we can't remove things
             // from their definitions). This might technically break our guarantees, but it works

--- a/src/main/java/org/checkerframework/specimin/JavaLangUtils.java
+++ b/src/main/java/org/checkerframework/specimin/JavaLangUtils.java
@@ -288,8 +288,8 @@ public final class JavaLangUtils {
     // TODO: can we get a list of such packages from the JDK instead of using this relatively-coarse
     // heuristic?
     return qualifiedName.startsWith("java.")
-        || qualifiedName.startsWith("javax.")
-        || qualifiedName.startsWith("com.sun.")
-        || qualifiedName.startsWith("jdk.");
+            || qualifiedName.startsWith("javax.")
+            || qualifiedName.startsWith("com.sun.")
+            || qualifiedName.startsWith("jdk.");
   }
 }

--- a/src/main/java/org/checkerframework/specimin/JavaLangUtils.java
+++ b/src/main/java/org/checkerframework/specimin/JavaLangUtils.java
@@ -277,4 +277,19 @@ public final class JavaLangUtils {
     boolean type2IsClass = type2.startsWith("Class<") || type2.startsWith("java.lang.Class<");
     return type1IsClass && type2IsClass;
   }
+
+  /**
+   * Is the given package name or fully-qualified name in one of the packages provided by the JDK?
+   *
+   * @param qualifiedName a package name or fully-qualified name of a class or interface
+   * @return true if qualifiedName is from the JDK
+   */
+  public static boolean inJdkPackage(String qualifiedName) {
+    // TODO: can we get a list of such packages from the JDK instead of using this relatively-coarse
+    // heuristic?
+    return qualifiedName.startsWith("java.")
+        || qualifiedName.startsWith("javax.")
+        || qualifiedName.startsWith("com.sun.")
+        || qualifiedName.startsWith("jdk.");
+  }
 }

--- a/src/main/java/org/checkerframework/specimin/JdkTypeSolver.java
+++ b/src/main/java/org/checkerframework/specimin/JdkTypeSolver.java
@@ -1,0 +1,16 @@
+package org.checkerframework.specimin;
+
+import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;
+
+/**
+ * A specialized version of Java Parser's ReflectionTypeSolver that allows access to compiler
+ * internals (e.g., com.sun and jdk. packages) in addition to the JRE, but does not include anything
+ * else on Specimin's classpath (especially JavaParser itself).
+ */
+public class JdkTypeSolver extends ReflectionTypeSolver {
+
+  @Override
+  protected boolean filterName(String name) {
+    return JavaLangUtils.inJdkPackage(name);
+  }
+}

--- a/src/main/java/org/checkerframework/specimin/MustImplementMethodsVisitor.java
+++ b/src/main/java/org/checkerframework/specimin/MustImplementMethodsVisitor.java
@@ -141,7 +141,7 @@ public class MustImplementMethodsVisitor extends ModifierVisitor<Void> {
     String methodSignature = method.getQualifiedSignature();
     // These classes are beyond our control. It's better to retain the implementations of all
     // abstract methods to ensure the code remains compilable.
-    if (methodSignature.startsWith("java.")) {
+    if (JavaLangUtils.inJdkPackage(methodSignature)) {
       return true;
     }
     return usedMembers.contains(methodSignature);

--- a/src/main/java/org/checkerframework/specimin/PrunerVisitor.java
+++ b/src/main/java/org/checkerframework/specimin/PrunerVisitor.java
@@ -229,7 +229,7 @@ public class PrunerVisitor extends ModifierVisitor<Void> {
           iterator.remove();
         }
         // all unresolvable interfaces that need to be removed belong to the Java package.
-        if (!typeFullName.startsWith("java.")) {
+        if (!JavaLangUtils.inJdkPackage(typeFullName)) {
           continue;
         }
         for (String classNeedInterfaceRemoved : classAndUnresolvedInterface.keySet()) {

--- a/src/main/java/org/checkerframework/specimin/SpeciminRunner.java
+++ b/src/main/java/org/checkerframework/specimin/SpeciminRunner.java
@@ -514,7 +514,7 @@ public class SpeciminRunner {
     // Set up the parser's symbol solver, so that we can resolve definitions.
     CombinedTypeSolver typeSolver =
         new CombinedTypeSolver(
-            new ReflectionTypeSolver(), new JavaParserTypeSolver(new File(root)));
+            new ReflectionTypeSolver(false), new JavaParserTypeSolver(new File(root)));
     for (String path : jarPaths) {
       typeSolver.add(new JarTypeSolver(path));
     }

--- a/src/main/java/org/checkerframework/specimin/SpeciminRunner.java
+++ b/src/main/java/org/checkerframework/specimin/SpeciminRunner.java
@@ -13,7 +13,6 @@ import com.github.javaparser.symbolsolver.JavaSymbolSolver;
 import com.github.javaparser.symbolsolver.resolution.typesolvers.CombinedTypeSolver;
 import com.github.javaparser.symbolsolver.resolution.typesolvers.JarTypeSolver;
 import com.github.javaparser.symbolsolver.resolution.typesolvers.JavaParserTypeSolver;
-import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;
 import com.github.javaparser.utils.SourceRoot;
 import java.io.File;
 import java.io.IOException;
@@ -513,8 +512,7 @@ public class SpeciminRunner {
   private static void updateStaticSolver(String root, List<String> jarPaths) throws IOException {
     // Set up the parser's symbol solver, so that we can resolve definitions.
     CombinedTypeSolver typeSolver =
-        new CombinedTypeSolver(
-            new ReflectionTypeSolver(false), new JavaParserTypeSolver(new File(root)));
+        new CombinedTypeSolver(new JdkTypeSolver(), new JavaParserTypeSolver(new File(root)));
     for (String path : jarPaths) {
       typeSolver.add(new JarTypeSolver(path));
     }

--- a/src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java
+++ b/src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java
@@ -2530,7 +2530,7 @@ public class UnsolvedSymbolVisitor extends ModifierVisitor<Void> {
     if (classfileIsInOriginalCodebase(qualifiedName)) {
       return;
     }
-    if (qualifiedName.startsWith("java.")) {
+    if (JavaLangUtils.inJdkPackage(qualifiedName)) {
       return;
     }
 

--- a/src/test/java/org/checkerframework/specimin/JdkCompilerTest.java
+++ b/src/test/java/org/checkerframework/specimin/JdkCompilerTest.java
@@ -1,0 +1,18 @@
+package org.checkerframework.specimin;
+
+import java.io.IOException;
+import org.junit.Test;
+
+/**
+ * This test checks that code that imports compiler components doesn't cause Specimin to fail. This
+ * is the test for https://github.com/njit-jerse/specimin/issues/284.
+ */
+public class JdkCompilerTest {
+  @Test
+  public void runTest() throws IOException {
+    SpeciminTestExecutor.runTestWithoutJarPaths(
+        "jdkcompiler",
+        new String[] {"com/example/Simple.java"},
+        new String[] {"com.example.Simple#bar(Tree)"});
+  }
+}

--- a/src/test/resources/jdkcompiler/expected/com/example/Simple.java
+++ b/src/test/resources/jdkcompiler/expected/com/example/Simple.java
@@ -1,0 +1,10 @@
+package com.example;
+
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.Tree.Kind;
+
+class Simple {
+    void bar(Tree tree) {
+        Kind kind = tree.getKind();
+    }
+}

--- a/src/test/resources/jdkcompiler/input/com/example/Simple.java
+++ b/src/test/resources/jdkcompiler/input/com/example/Simple.java
@@ -1,0 +1,11 @@
+package com.example;
+
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.Tree.Kind;
+
+class Simple {
+    // Target method.
+    void bar(Tree tree) {
+        Kind kind = tree.getKind();
+    }
+}

